File: run_inf_kfold.py
#!/usr/bin/env python3
"""
K-fold FF ONNX inference

Usage:
  python run_inf_kfold.py --model-dir models
"""

import argparse
import json
import logging
from pathlib import Path
from typing import Dict, List

import numpy as np
import onnxruntime as ort

LOG = logging.getLogger("FF_KFOLD")


def setup_logging(verbose: bool) -> None:
    level = logging.DEBUG if verbose else logging.INFO
    logging.basicConfig(level=level, format="%(asctime)s [%(levelname)s] %(message)s")


class KFoldFFONNXRunner:
    def __init__(self, model_dir: str):
        self.model_dir = Path(model_dir)

        folds: List[int] = []
        for path in self.model_dir.glob("feature_order_fold*.json"):
            stem = path.stem 
            if "fold" not in stem:
                continue
            try:
                idx = int(stem.split("fold", 1)[1])
                folds.append(idx)
            except ValueError:
                continue

        if not folds:
            raise RuntimeError(f"No feature_order_fold*.json found in {self.model_dir}")

        self.n_folds = max(folds) + 1
        LOG.info("Initialising K-fold runner from %s (n_folds=%d)", self.model_dir, self.n_folds)

        self.sessions: List[ort.InferenceSession] = []
        self.fold_feature_order: List[List[str]] = []
        self.fold_feature_index: List[Dict[str, int]] = []
        self.fold_dm_values: List[List[int]] = []
        self.fold_dm_indices: List[List[int]] = []

        for fold in range(self.n_folds):
            onnx_path = self.model_dir / f"model_fold{fold}.onnx"
            json_path = self.model_dir / f"feature_order_fold{fold}.json"

            if not onnx_path.exists():
                raise FileNotFoundError(f"Missing ONNX model: {onnx_path}")
            if not json_path.exists():
                raise FileNotFoundError(f"Missing feature-order JSON: {json_path}")

            data = json.loads(json_path.read_text())
            feature_order = list(data["feature_order"])
            self.fold_feature_order.append(feature_order)

            feature_index = {name: i for i, name in enumerate(feature_order)}
            self.fold_feature_index.append(feature_index)

            dm_vals: List[int] = []
            dm_idx: List[int] = []
            for i, name in enumerate(feature_order):
                if name.startswith("decayMode_"):
                    dm_vals.append(int(name.split("_", 1)[1]))
                    dm_idx.append(i)
            self.fold_dm_values.append(dm_vals)
            self.fold_dm_indices.append(dm_idx)

            sess = ort.InferenceSession(str(onnx_path), providers=["CPUExecutionProvider"])
            self.sessions.append(sess)

            LOG.info("Fold %d: ONNX initialised from %s", fold, onnx_path)

        base_order = self.fold_feature_order[0]
        self.scalar_feature_names = [n for n in base_order if not n.startswith("decayMode_")]
        self.decay_feature_names = [n for n in base_order if n.startswith("decayMode_")]

        LOG.info("Feature order (fold 0): %s", ", ".join(base_order))

    def compute_w_ff(self, event_id: np.ndarray, **feat_arrays: np.ndarray) -> np.ndarray:
        event_id = np.asarray(event_id, dtype=np.int64).ravel()
        n = event_id.size
        if n == 0:
            return np.zeros(0, dtype=np.float32)

        if "decayMode" not in feat_arrays:
            raise ValueError("Missing required feature 'decayMode'")

        decay_mode = np.asarray(feat_arrays["decayMode"], dtype=np.int32).ravel()
        if decay_mode.size != n:
            raise ValueError("decayMode length mismatch with event_id")

        LOG.info("compute_w_ff: n_taus=%d", n)
        LOG.debug("User features: %s", ", ".join(sorted(feat_arrays.keys())))
        LOG.debug("Scalar features: %s", ", ".join(self.scalar_feature_names))
        LOG.debug("Decay one-hot features: %s", ", ".join(self.decay_feature_names))

        scalar_arrays: Dict[str, np.ndarray] = {}
        for name in self.scalar_feature_names:
            if name not in feat_arrays:
                raise ValueError(f"Missing required feature '{name}'")
            arr = np.asarray(feat_arrays[name], dtype=np.float32).ravel()
            if arr.size != n:
                raise ValueError(f"Feature '{name}' has length {arr.size}, expected {n}")
            scalar_arrays[name] = arr

        w = np.zeros(n, dtype=np.float32)

        for fold in range(self.n_folds):
            mask = (event_id % self.n_folds) == fold
            idxs = np.nonzero(mask)[0]
            if idxs.size == 0:
                continue

            fo = self.fold_feature_order[fold]
            fi = self.fold_feature_index[fold]
            dm_vals = self.fold_dm_values[fold]
            dm_idx = self.fold_dm_indices[fold]

            m = idxs.size
            n_feat = len(fo)
            x = np.zeros((m, n_feat), dtype=np.float32)

            for name, arr in scalar_arrays.items():
                j = fi.get(name)
                if j is None:
                    continue
                x[:, j] = arr[idxs]

            # decayMode one-hot
            dm_slice = decay_mode[idxs]
            for dv, col in zip(dm_vals, dm_idx):
                x[:, col] = (dm_slice == dv).astype(np.float32)

            sess = self.sessions[fold]
            y = sess.run(["w_ff"], {"raw_input": x})[0].reshape(-1).astype(np.float32)
            w[idxs] = y

        return w


def main() -> None:
    ap = argparse.ArgumentParser("K-fold FF ONNX inference")
    ap.add_argument(
        "--model-dir",
        required=True,
        help="Directory with model_fold*.onnx and feature_order_fold*.json",
    )
    ap.add_argument("-v", "--verbose", action="store_true", help="Verbose logging")
    args = ap.parse_args()

    setup_logging(args.verbose)

    runner = KFoldFFONNXRunner(args.model_dir)

    # example
    n = 8
    event_id   = np.array([1001, 1001, 1002, 1003, 1004, 1005, 1005, 1006], dtype=np.int64)
    decay_mode = np.array([0, 1, 2, 10, 11, 0, 1, 2], dtype=np.int32)

    def fill(v: float) -> np.ndarray:
        return np.full(n, v, dtype=np.float32)

    pt             = fill(45.0)
    eta            = fill(0.3)
    mass           = fill(1.2)
    seedingJet_pt  = fill(50.0)
    seedingJet_eta = fill(0.1)
    seedingJet_mass= fill(10.0)
    btagPNetB      = fill(0.2)
    btagPNetCvB    = fill(0.1)
    btagPNetCvL    = fill(0.4)
    btagPNetCvNotB = fill(0.3)
    btagPNetQvG    = fill(0.5)

    w = runner.compute_w_ff(
        event_id=event_id,
        decayMode=decay_mode,
        pt=pt,
        eta=eta,
        mass=mass,
        seedingJet_pt=seedingJet_pt,
        seedingJet_eta=seedingJet_eta,
        seedingJet_mass=seedingJet_mass,
        btagPNetB=btagPNetB,
        btagPNetCvB=btagPNetCvB,
        btagPNetCvL=btagPNetCvL,
        btagPNetCvNotB=btagPNetCvNotB,
        btagPNetQvG=btagPNetQvG,
    )

    print("event_id  fold  decayMode  w_ff")
    for i in range(n):
        fold = int(event_id[i] % runner.n_folds)
        print(f"{int(event_id[i]):7d}  {fold:4d}  {int(decay_mode[i]):9d}  {w[i]:.6g}")


if __name__ == "__main__":
    main()

File: run_inf_kfold.cc
// K-fold FF ONNX inference
//
// Usage:
//   g++ run_inf_kfold.cc -o ff_infer \
//       -I/cvmfs/sft.cern.ch/lcg/views/LCG_107/x86_64-el9-gcc11-opt/include/onnxruntime \
//       -L/cvmfs/sft.cern.ch/lcg/views/LCG_107/x86_64-el9-gcc11-opt/lib64 \
//       -lonnxruntime -std=c++17 -O2
//
//   ./ff_infer models 5

#include "onnxruntime_cxx_api.h"
#include <nlohmann/json.hpp>

#include <cmath>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <map>
#include <memory>
#include <stdexcept>
#include <string>
#include <vector>

using json = nlohmann::json;

struct FoldConfig {
    std::vector<std::string> feature_order;
    std::map<std::string, std::size_t> feature_index;
    std::vector<int> decay_modes;
    std::vector<std::size_t> decay_indices;

    explicit FoldConfig(const std::vector<std::string>& fo)
        : feature_order(fo)
    {
        for (std::size_t i = 0; i < feature_order.size(); ++i) {
            const auto& name = feature_order[i];
            feature_index[name] = i;
            if (name.rfind("decayMode_", 0) == 0) {
                int dm = std::stoi(name.substr(std::string("decayMode_").size()));
                decay_modes.push_back(dm);
                decay_indices.push_back(i);
            }
        }
    }
};

class KFoldFFONNX {
public:
    KFoldFFONNX(const std::string& model_dir, int n_folds)
        : env_(ORT_LOGGING_LEVEL_WARNING, "FF_KFold"),
          n_folds_(n_folds)
    {
        if (n_folds_ <= 0) {
            throw std::runtime_error("n_folds must be > 0");
        }

        std::cout << "[info] Initialising KFoldFFONNX from " << model_dir
                  << " with n_folds=" << n_folds_ << "\n";

        for (int f = 0; f < n_folds_; ++f) {
            std::string onnx_path = model_dir + "/model_fold" + std::to_string(f) + ".onnx";
            std::string cfg_path  = model_dir + "/feature_order_fold" + std::to_string(f) + ".json";

            std::ifstream cf(cfg_path);
            if (!cf) {
                throw std::runtime_error("Cannot open feature-order JSON: " + cfg_path);
            }
            json j;
            cf >> j;
            std::vector<std::string> feature_order =
                j.at("feature_order").get<std::vector<std::string>>();

            fold_cfgs_.emplace_back(feature_order);

            std::ifstream mf(onnx_path);
            if (!mf) {
                throw std::runtime_error("Cannot open ONNX file: " + onnx_path);
            }

            Ort::SessionOptions opts;
            opts.SetIntraOpNumThreads(1);
            opts.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_EXTENDED);
            sessions_.emplace_back(
                std::make_unique<Ort::Session>(env_, onnx_path.c_str(), opts)
            );

            std::cout << "[info] Fold " << f << ": ONNX initialised from " << onnx_path << "\n";
        }

        const auto& fo0 = fold_cfgs_.front().feature_order;
        for (const auto& name : fo0) {
            if (name.rfind("decayMode_", 0) != 0) {
                scalar_features_.push_back(name);
            }
        }

        std::cout << "[info] Feature order (fold 0): ";
        for (std::size_t i = 0; i < fo0.size(); ++i) {
            std::cout << fo0[i];
            if (i + 1 < fo0.size()) std::cout << ", ";
        }
        std::cout << "\n";
    }

    std::vector<float> compute_w_ff(
        const std::vector<long long>& event_id,
        const std::map<std::string, std::vector<float>>& features)
    {
        std::size_t n = event_id.size();
        if (n == 0) {
            return {};
        }

        std::cout << "[info] compute_w_ff: n_events=" << n << "\n";

        auto dm_it = features.find("decayMode");
        if (dm_it == features.end()) {
            throw std::runtime_error("Missing feature 'decayMode'.");
        }
        if (dm_it->second.size() != n) {
            throw std::runtime_error("decayMode feature size mismatch.");
        }

        std::vector<int> decay_mode(n);
        for (std::size_t i = 0; i < n; ++i) {
            decay_mode[i] = static_cast<int>(std::lround(dm_it->second[i]));
        }

        for (const auto& name : scalar_features_) {
            auto it = features.find(name);
            if (it == features.end()) {
                throw std::runtime_error("Missing scalar feature: " + name);
            }
            if (it->second.size() != n) {
                throw std::runtime_error("Size mismatch for feature: " + name);
            }
        }

        std::vector<float> out(n, 0.f);

        for (int f = 0; f < n_folds_; ++f) {
            std::vector<std::size_t> idxs;
            idxs.reserve(n);
            for (std::size_t i = 0; i < n; ++i) {
                if (static_cast<int>(event_id[i] % n_folds_) == f) {
                    idxs.push_back(i);
                }
            }
            if (idxs.empty()) continue;

            const auto& cfg  = fold_cfgs_[f];
            auto& sess       = *sessions_[f];

            std::size_t m      = idxs.size();
            std::size_t n_feat = cfg.feature_order.size();
            std::vector<float> input_data(m * n_feat, 0.f);

            for (const auto& name : scalar_features_) {
                const auto& col = features.at(name);
                auto fi_it = cfg.feature_index.find(name);
                if (fi_it == cfg.feature_index.end()) continue;
                std::size_t j = fi_it->second;
                for (std::size_t k = 0; k < m; ++k) {
                    std::size_t i = idxs[k];
                    input_data[k * n_feat + j] = col[i];
                }
            }

            // decayMode one-hot
            for (std::size_t dm_i = 0; dm_i < cfg.decay_modes.size(); ++dm_i) {
                int dm_val        = cfg.decay_modes[dm_i];
                std::size_t col_j = cfg.decay_indices[dm_i];
                for (std::size_t k = 0; k < m; ++k) {
                    std::size_t i = idxs[k];
                    input_data[k * n_feat + col_j] =
                        (decay_mode[i] == dm_val) ? 1.f : 0.f;
                }
            }

            Ort::MemoryInfo mem_info = Ort::MemoryInfo::CreateCpu(
                OrtArenaAllocator, OrtMemTypeDefault);
            std::vector<int64_t> shape{
                static_cast<int64_t>(m),
                static_cast<int64_t>(n_feat)
            };

            Ort::Value input_tensor = Ort::Value::CreateTensor<float>(
                mem_info,
                input_data.data(),
                input_data.size(),
                shape.data(),
                shape.size()
            );

            const char* input_names[]  = {"raw_input"};
            const char* output_names[] = {"w_ff"};

            auto output_tensors = sess.Run(
                Ort::RunOptions{nullptr},
                input_names, &input_tensor, 1,
                output_names, 1
            );

            float* out_data = output_tensors[0].GetTensorMutableData<float>();
            for (std::size_t k = 0; k < m; ++k) {
                out[idxs[k]] = out_data[k];
            }
        }

        return out;
    }

private:
    Ort::Env env_;
    int n_folds_;
    std::vector<std::unique_ptr<Ort::Session>> sessions_;
    std::vector<FoldConfig> fold_cfgs_;
    std::vector<std::string> scalar_features_;
};

int main(int argc, char* argv[]) {
    if (argc < 3) {
        std::cerr << "Usage: " << argv[0] << " MODEL_DIR N_FOLDS\n";
        return 1;
    }

    std::string model_dir = argv[1];
    int n_folds = std::stoi(argv[2]);

    try {
        KFoldFFONNX kff(model_dir, n_folds);

        // example
        std::size_t n = 8;
        std::vector<long long> event_id{1000, 1000, 1002, 1003, 1004, 1005, 1006, 1006};

        auto fill = [n](float v) {
            return std::vector<float>(n, v);
        };

        std::map<std::string, std::vector<float>> feats;
        feats["decayMode"]      = {0.f, 1.f, 2.f, 10.f, 11.f, 0.f, 1.f, 2.f};
        feats["pt"]             = fill(45.0f);
        feats["eta"]            = fill(0.3f);
        feats["mass"]           = fill(1.2f);
        feats["seedingJet_pt"]  = fill(50.0f);
        feats["seedingJet_eta"] = fill(0.1f);
        feats["seedingJet_mass"]= fill(10.0f);
        feats["btagPNetB"]      = fill(0.2f);
        feats["btagPNetCvB"]    = fill(0.1f);
        feats["btagPNetCvL"]    = fill(0.4f);
        feats["btagPNetCvNotB"] = fill(0.3f);
        feats["btagPNetQvG"]    = fill(0.5f);

        auto w = kff.compute_w_ff(event_id, feats);

        std::cout << "event_id  fold  decayMode  w_ff\n";
        for (std::size_t i = 0; i < n; ++i) {
            int fold = static_cast<int>(event_id[i] % n_folds);
            int dm   = static_cast<int>(std::lround(feats["decayMode"][i]));
            std::cout << event_id[i] << "  "
                      << fold << "  "
                      << dm << "  "
                      << w[i] << "\n";
        }

        return 0;
    } catch (const std::exception& ex) {
        std::cerr << "[error] " << ex.what() << "\n";
        return 1;
    }
}